src/server/main.cpp

```diff
@@ -21,13 +30,67 @@ const int SIZE = 2;
 
 uint16_t *shared_mem;
 
+void doUpdate(const SwtFB &fb, const swtfb_update &s) {
+  auto mxcfb_update = s.update;
+  auto rect = mxcfb_update.update_region;
+
+#ifdef DEBUG_DIRTY
+  std::cerr << "Dirty Region: " << rect.left << " " << rect.top << " "
+            << rect.width << " " << rect.height << endl;
+#endif
+
+  // For now, we are using two waveform modes:
+  // 0: init (same as GL16)
+  // 1: DU - direct update, fast
+  // 2: GC16 - high fidelity (slow)
+  // 3: GL16 - what the rm is using
+  // 8: highlight (same as high fidelity)
+
+  int waveform = mxcfb_update.waveform_mode;
+  // full = 1, partial = 0
+  auto update_mode = mxcfb_update.update_mode;
+
+  auto flags = update_mode & 0x1;
+
+  // TODO: Get sync from client (wait for second ioctl? or look at stack?)
+  // There are only two occasions when the original rm1 library sets sync to
+  // true. Currently we detect them by the other data. Ideally we should
+  // correctly handle the corresponding ioctl (empty rect and flags == 2?).
+  if (waveform == /*init*/ 0 && update_mode == /* full */ 1) {
+    flags |= 2;
+    std::cerr << "SERVER: sync" << std::endl;
+  } else if (rect.left == 0 && rect.top > 1800 &&
+             waveform == /* grayscale */ 3 && update_mode == /* full */ 1) {
+    std::cerr << "server sync, x2: " << rect.width << " y2: " << rect.height
+              << std::endl;
+    flags |= 2;
+  }
+
+#ifdef DEBUG
+  std::cerr << "doUpdate " << endl;
+  cerr << "mxc: waveform_mode " << mxcfb_update.waveform_mode << endl;
+  cerr << "mxc: update mode " << mxcfb_update.update_mode << endl;
+  cerr << "mxc: update marker " << mxcfb_update.update_marker << endl;
+  cerr << "final: waveform " << waveform;
+  cerr << " flags " << flags << endl << endl;
+#endif
+
+  fb.DrawRaw(rect.left, rect.top, rect.width, rect.height, waveform, flags);
+
+#ifdef DEBUG_TIMING
+  cerr << get_now() -.ms << "ms E2E " << rect.width << " " << rect.height
```

Comment:
not sure this will compile? (what is `get_now()-.ms` do?)

Reply:
Just copied, I removed it for now.